### ISSUE TITLE
[bug] event emitters cause uncaught errors

### ISSUE THREAD

**Issue by nicholaschiang** [PLhery/node-twitter-api-v2#229]
**Describe the bug**

Because of some event emitter that either `twitter-api-v2` is using or a dependency of `twitter-api-v2`, errors are thrown twice: once that can be caught within an `async` function but then again when they are emitted by that `EventEmitter` and cannot be caught, resulting in an uncaught error and the Node.js process terminating the an error code (1).

**To Reproduce**
Please indicate all steps that lead to this bug:
1. Send an async API request that causes an error (e.g. "429 Too Many Requests").
2. Wrap that API request in a `try`, `catch` clause and log the caught error (but don't re-throw).
3. Notice that the error is correctly logged but it is still re-emitted later in an uncaught exception:

```text
[dev:remix] [info] Fetching followed lists for user (1329661759020363778) lists...
[dev:remix] [info] Fetching owned lists for user (1329661759020363778) lists...
[dev:remix] [error] You just hit the rate limit! Limit for this endpoint is 15 requests!
[dev:remix] [error] Request counter will reset at timestamp 3/21/2022, 9:13:38 PM.
[dev:remix] Error: You defined an action for route "routes/sync/lists" but didn't return anything from your `action` function. Please return a value or `null`.
[dev:remix]     at Object.callRouteAction (/home/nchiang/repos/tweetscape/node_modules/@remix-run/server-runtime/data.js:54:11)
[dev:remix]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[dev:remix]     at handleDataRequest (/home/nchiang/repos/tweetscape/node_modules/@remix-run/server-runtime/server.js:110:18)
[dev:remix]     at requestHandler (/home/nchiang/repos/tweetscape/node_modules/@remix-run/server-runtime/server.js:45:20)
[dev:remix]     at /home/nchiang/repos/tweetscape/node_modules/@remix-run/express/server.js:43:22
[dev:remix] PATCH /sync/lists?_data=routes%2Fsync%2Flists 500 - - 266.739 ms
[dev:remix] Error: Request failed with code 429
[dev:remix]     at RequestHandlerHelper.createResponseError (/home/nchiang/repos/tweetscape/node_modules/twitter-api-v2/dist/client-mixins/request-handler.helper.js:99:16)
[dev:remix]     at RequestHandlerHelper.onResponseEndHandler (/home/nchiang/repos/tweetscape/node_modules/twitter-api-v2/dist/client-mixins/request-handler.helper.js:248:25)
[dev:remix]     at Gunzip.emit (node:events:520:28)
[dev:remix]     at endReadableNT (node:internal/streams/readable:1346:12)
[dev:remix]     at processTicksAndRejections (node:internal/process/task_queues:83:21)
[dev:remix] yarn run dev:remix exited with code 1
```

That :point_up: is the result of the following code (I've omitted lines for brevity):

```typescript
export const action: ActionFunction = async ({ request }) => {
  try {
    const api = new TwitterApi(accessToken);
    log.info(`Fetching followed lists for ${context}...`);
    const res = await api.v2.listFollowed(uid);
  } catch (e) {
    if (e instanceof ApiResponseError && e.rateLimitError && e.rateLimit) {
      const msg1 = 
        `You just hit the rate limit! Limit for this endpoint is ` +
        `${e.rateLimit.limit} requests!`
      const msg2 =
        `Request counter will reset at timestamp ` +
        `${new Date(e.rateLimit.reset * 1000).toLocaleString()}.`
      log.error(msg1); // This is logged.
      log.error(msg2); // This is also logged, meaning the error was caught.
    }
  }
};
```

**Expected behavior**

Errors should only be thrown once and, once caught, shouldn't result in a non-zero Node.js process exit.

**Version**
 - Node.js version: 16.14.0
 - Lib version: 1.11.1
 - OS (especially if you use Windows): Pop_OS! 20.04 (Ubuntu 20.04 LTS)

**Additional context**

This may have something to do [with this](https://stackoverflow.com/questions/64568569/how-to-catch-an-error-on-an-eventemitter-that-runs-an-async-function-on-emit).

**Comment by alkihis** [comment]
Hi,

Even by using the same endpoint wrapper as you, I cannot reproduce this issue.
Error is catched correctly on my side and my program does not crash.

I tried to replicate your issue and get a 429 after 15+ tries, but it doesn't crash either.

Plus, what you're seeing in the crash dump `at RequestHandlerHelper.createResponseError` means that what's making your program crash is the error given to `reject` function of `Promise` executor:

```ts
// Handle bad error codes
const code = this.res.statusCode!;
if (code >= 400) {
  reject(this.createResponseError({ data, res: this.res, rateLimit, code }));
  return;
}
```

It cannot be an uncatched `error` event, because that causes the error to come from elsewhere in stack trace.
As a `Promise` cannot be rejected or resolved twice, it means another call (maybe in another piece of code?) isn't catched properly.

**Comment by nicholaschiang** [comment]
Alright, I found the error! It has to do with using a plugin:

```javascript
import { TwitterApiRateLimitPlugin } from '@twitter-api-v2/plugin-rate-limit';
import { TwitterApi } from 'twitter-api-v2';
import express from 'express';

function doSomethingAsync() {
  return new Promise((res, rej) => setTimeout(res, 500));
}

async function getTwitterClient() {
  await doSomethingAsync();
  // Get rid of this plugin and the error catches as expected.
  const plugin = new TwitterApiRateLimitPlugin();
  return new TwitterApi('INVALID_TOKEN', { plugins: [plugin] });
}

const app = express();
const port = 3001;

app.get('/error', async (req, res) => {
  try {
    await doSomethingAsync();
    const api = await getTwitterClient();
    await Promise.all(Array(5).fill(null).map(async () => {
      await doSomethingAsync();
      await api.v2.listTweets('INVALID_LIST_ID');
    }));
    res.status(200).send('Success');
  } catch (e) {
    // When the plugin is used, this still catches but an error is emitted
    // down the line causing the Node.js process to exit with code 1.
    console.error('Caught error:', e.stack);
    res.status(500).send('Error');
  }
});

app.listen(port, () => console.log(`API listening on port ${port}`));
```

I also created a Remix repro here: https://github.com/nicholaschiang/errors

You should be able to simply clone it, `yarn install`, and then `node app.mjs` to see the Express server error.

**Comment by alkihis** [comment]
Nice catch!
I see what's happening on my side now. Thanks for your repo, I will explore it too.

**Comment by alkihis** [comment]
Should be fixed in 1.11.2.

**Comment by nicholaschiang** [comment]
Love it! Thanks so much @alkihis!
